Service maturity docs (#128)
port-labs/port-docs: docs/software-catalog/scorecard/scorecard.md

---
sidebar_position: 1
---

# Scorecard

## What is a Scorecard?

> **Scorecards** enable us to create a set of rules that will determine our Port entity's level based on its properties.
> Each scorecard has a set of rules that affects its total level, a rule has a `level` property which is one of the following: `Gold`, `Silver` or `Bronze`. Each rule has specific conditions and when they pass the scorecard level increases.

Reviewer: ```suggestion
**Scorecards** enable us to create a set of rules that will determine the level of our Port Entities, based on their properties.
Each scorecard has a set of rules that affects its total level, a rule has a `level` property which is one of the following: `Gold`, `Silver` or `Bronze`. Each rule has specific conditions, and the scorecard level increases when they pass.
```

> **For example**, to keep track of your organization's `Services` maturity, we can create a set of scorecards on top of a `Service` [Blueprint](../blueprint/blueprint.md) that will keep track of their progress, here are some scorecards that we can set:

Reviewer: ```suggestion
**For example**, to keep track of your organization's `Services` maturity, we can create a set of scorecards on top of a `Service` [Blueprint](../blueprint/blueprint.md) that will keep track of their progress. here are some scorecards that we can set:
```

- Ownership
  - Has a defined on-call?
  - Has a team?
- Security
  - Does it's Snyk vulnerabilities < 1
- Production Readiness
  - Using a Monitoring service?
- Development Quality
  - Have a linter configured?
  - Have tests?

> In the end, within the specific entity page we will get a tab per scorecard which shows the progress and the rules status.

Reviewer: ```suggestion
In the end, within the specific Entity profile, we will get a tab per scorecard that shows the rules' progress and status.
```

| `query`      | `Object` | The query is built from an array of [conditions](#condition-structure-table) and a `combinator` (or / and) that will define the |

## Condition structure table

| Field      | Description                                                                                                                                                                                                                          |
| ---------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `operator` | Search operator to use when evaluating this rule, for example `=` `!=` `contains` `doesNotContains` `isEmpty` `isNotEmpty` below                                                                                                     |
| `property` | Property to filter by according to its value. It can be a [meta-property](../../software-catalog/blueprint/mirror-properties.md#meta-property-mirror-property) such as `$identifier`, or a standard property such as `slack_channel` |

> | `value`    | The value to compare to (not required in isEmpty and isNotEmpty operators)                                                                                                                                                           |

Reviewer: ```suggestion
| `value`    | Value to compare to (not required in isEmpty and isNotEmpty operators)                                                                                                                                                           |
```

## Scorecard total level calculation

As shown above, a Scorecard is built from several rules, and each one of them has a `level` property.

> The available levels are `Gold`, `Silver`, `Bronze`, if an entity passes all of the rules from a certain level it gets to that level, while ensuring it passed all previous levels.

Reviewer: ```suggestion
The available levels are `Gold`, `Silver`, `Bronze`. Once an Entity passes the conditions for a certain level, its level changes accordingly.
```

:::note

> if entity didn't pass all of the rules of the `Bronze` level it will be in a `Basic` Tier.

Reviewer: ```suggestion
If an Entity hasn't passed the rules determining a `Bronze` level, it will be a `Basic` Tier.
```

:::

## Scorecard example

Please see the following example of an ownership scorecard. It has two rules, one for checking if a defined on-call exists and another for checking if a team exists. The conditions for each rule use the "isNotEmpty" operator to check if the specified property is not empty.

```json showLineNumbers
{
  "title": "Ownership",
  "identifier": "ownership",
  "rules": [
    {
      "title": "Has a defined on call?",
      "identifier": "has_on_call",
      "level": "Gold",
      "query": {
        "combinator": "and",
        "conditions": [
          {
            "operator": "isNotEmpty",
            "property": "on_call"
          }
        ]
      }
    },
    {
      "title": "Has a team?",
      "identifier": "has_team",
      "level": "Silver",
      "query": {
        "combinator": "and",
        "conditions": [
          {
            "operator": "isNotEmpty",
            "property": "$team"
          }
        ]
      }
    }
  ]
}
```

> ## Next Steps

Reviewer: ```suggestion
## Next steps
```
